remove empty "implements" (#2695)
TYPO3-Documentation/TYPO3CMS-Reference-CoreApi: Documentation/ApiOverview/Logging/Writers/Index.rst

Usage in Custom Class
---------------------

All log writers can be used in your own classes. If the service is configured to use autowiring 

> you can initialize the loggers in the __construct method of your class :code:`\MyDomain\MyExtension\MyFolder\MyClass`) . 

Reviewer: ```suggestion
you can inject a logger into the :php:`__construct` method of your class :php:`\MyVendor\MyExtension\MyFolder\MyClass`) since TYPO3 v11 LTS. 
```

I would prefer "inject" instead of initialize. Also we use "MyVendor" as vendor example throughout the documentation.

> If no autowiring is used the service class however must implement the interface :code:`\Psr\Log\LoggerAwareInterface` and :code:`\Psr\Log\LoggerAwareTrait` and use no constructor.

Reviewer: ```suggestion
If autowiring is disabled, the service class however must implement the interface :php:`\Psr\Log\LoggerAwareInterface` and use :php:`\Psr\Log\LoggerAwareTrait`.
```

"and use no constructor" is misleading, as you can use a constructor to inject additional dependencies.

I also suggest to add an example for this case.

Usage in Custom Class
---------------------

All log writers can be used in your own classes. If the service is configured to use autowiring 
you can inject a logger into the :php:`__construct` method of your class :php:`\MyVendor\MyExtension\MyFolder\MyClass`) since TYPO3 v11 LTS.

.. code-block:: php

    namespace MyDomain\MyExtension\MyFolder;

    use Psr\Log\LoggerInterface;

>     class MyClass implements SingletonInterface {

Reviewer: ```suggestion
    class MyClass {
```

       private LoggerInterface $logger;

       public function __construct(LoggerInterface $logger) {
           $this->logger = $logger;
       }
        ...
           $this->logger->info('My class is executed.');
           if ($error) {
              $this->logger->error('error in class MyClass');
           }
        ...
    }

If autowiring is disabled, the service class however must implement the interface :php:`\Psr\Log\LoggerAwareInterface` and use the :php:`\Psr\Log\LoggerAwareTrait`.

.. code-block:: php

    namespace MyVendor\MyExtension\MyFolder;

    use Psr\Log\LoggerAwareInterface;
    use Psr\Log\LoggerAwareTrait;

    class MyClass implements LoggerAwareInterface {
       use LoggerAwareTrait;

>         ...

Reviewer 2: Here a method is now missing

Reviewer: Yup, it was before also. Will do a follow-up to clean also other things up.

Author: Which method is missing?

Reviewer 3: We will do that in a follow up, @brotkrueml can tag you in the follow up

Usage in Custom Class
---------------------

All log writers can be used in your own classes. If the service is configured to use autowiring 
you can inject a logger into the :php:`__construct` method of your class :php:`\MyVendor\MyExtension\MyFolder\MyClass`) since TYPO3 v11 LTS.

.. code-block:: php

    namespace MyDomain\MyExtension\MyFolder;

    use Psr\Log\LoggerInterface;

>     use TYPO3\CMS\Core\SingletonInterface;

Reviewer: ```suggestion
```

    class MyClass implements SingletonInterface {
       private LoggerInterface $logger;

       public function __construct(LoggerInterface $logger) {
           $this->logger = $logger;
       }
        ...
           $this->logger->info('My class is executed.');
           if ($error) {
              $this->logger->error('error in class MyClass');
           }
        ...
    }

If autowiring is disabled, the service class however must implement the interface :php:`\Psr\Log\LoggerAwareInterface` and use the :php:`\Psr\Log\LoggerAwareTrait`.

.. code-block:: php

>     namespace MyDomain\MyExtension\MyFolder;

Reviewer: ```suggestion
    namespace MyVendor\MyExtension\MyFolder;
```

Usage in Custom Class
---------------------

All log writers can be used in your own classes. If the service is configured to use autowiring 
you can inject a logger into the :php:`__construct` method of your class :php:`\MyVendor\MyExtension\MyFolder\MyClass`) since TYPO3 v11 LTS.

.. code-block:: php

    namespace MyDomain\MyExtension\MyFolder;

    use Psr\Log\LoggerInterface;
    use TYPO3\CMS\Core\SingletonInterface;

    class MyClass implements SingletonInterface {
       private LoggerInterface $logger;

       public function __construct(LoggerInterface $logger) {
           $this->logger = $logger;
       }
        ...
           $this->logger->info('My class is executed.');
           if ($error) {
              $this->logger->error('error in class MyClass');
           }
        ...
    }

If autowiring is disabled, the service class however must implement the interface :php:`\Psr\Log\LoggerAwareInterface` and use the :php:`\Psr\Log\LoggerAwareTrait`.

.. code-block:: php

    namespace MyDomain\MyExtension\MyFolder;

    use Psr\Log\LoggerAwareInterface;
    use Psr\Log\LoggerAwareTrait;
    use TYPO3\CMS\Core\SingletonInterface;

>     class MyClass implements SingletonInterface, LoggerAwareInterface {

Reviewer: ```suggestion
    class MyClass implements LoggerAwareInterface {
```

Usage in Custom Class
---------------------

All log writers can be used in your own classes. If the service is configured to use autowiring 
you can initialize the loggers in the __construct method of your class :code:`\MyDomain\MyExtension\MyFolder\MyClass`) . 

> If no autowiring is used the service class however must implement the interface :code:`\Psr\Log\LoggerAwareInterface` and :code:`\Psr\Log\LoggerAwareTrait` and use no constructor.

Author: I have added a second example.

> If no autowiring is used the service class however must implement the interface :code:`\Psr\Log\LoggerAwareInterface` and :code:`\Psr\Log\LoggerAwareTrait` and use no constructor.

Author: > If no autowiring is used the service class however must implement the interface :code:`\Psr\Log\LoggerAwareInterface`, `use` the :code:`\Psr\Log\LoggerAwareTrait` and never use a constructor.

But there is an example in TYPO3 12 using TRAIT and a constructor.

    class ArrayAdapter implements AdapterInterface, CacheInterface, LoggerAwareInterface, ResettableInterface
    {
        use LoggerAwareTrait;
    
        /**
         * @param bool $storeSerialized Disabling serialization can lead to cache corruptions when storing mutable values but increases performance otherwise
         */
        public function __construct(int $defaultLifetime = 0, bool $storeSerialized = true, float $maxLifetime = 0, int $maxItems = 0)
        {

> If no autowiring is used the service class however must implement the interface :code:`\Psr\Log\LoggerAwareInterface` and :code:`\Psr\Log\LoggerAwareTrait` and use no constructor.

Reviewer 2: Currently I can't test it.
So if you write about an issue you should setup perhaps an instance to test.

> If no autowiring is used the service class however must implement the interface :code:`\Psr\Log\LoggerAwareInterface` and :code:`\Psr\Log\LoggerAwareTrait` and use no constructor.

Reviewer: The ArrayAdapter is part of symfony/cache component. So, Symfony handles it somehow. symfony/cache is not used by TYPO3, but defined as dependency from another dependency, therefore installed.

TYPO3 has a LoggerAwarePass which takes care to call the setLogger() method of the trait if the class implements the LoggerAwareInterface via DI.

Additionally, GeneralUtility::makeInstance() checks for the interface and calls then the setLogger() method.

From my POV:
- Instantiating a class with GeneralUtility::makeInstance() works in a case when scalar values are given via constructor (when also given to makeInstance()).
- DI works with static values in the constructor if they are given from Services.[yaml|php].

@DavidBruchmann Using LoggerAwareInterface and a constructor (for injecting other dependencies) works in v10 and v11, as I used them in extensions/projects. As mentioned above, it is just a compiler pass which does the magic, independent, of a constructor is available or not.

.. code-block:: php

    namespace MyDomain\MyExtension\MyFolder;

    use Psr\Log\LoggerInterface;

    class MyClass implements SingletonInterface {
       private LoggerInterface $logger;

       public function __construct(LoggerInterface $logger) {
           $this->logger = $logger;
       }
        ...
           $this->logger->info('My class is executed.');
           if ($error) {
              $this->logger->error('error in class MyClass');
           }
        ...
    }

If autowiring is disabled, the service class however must implement the interface :php:`\Psr\Log\LoggerAwareInterface` and use the :php:`\Psr\Log\LoggerAwareTrait`.

.. code-block:: php

    namespace MyDomain\MyExtension\MyFolder;

    use Psr\Log\LoggerAwareInterface;
    use Psr\Log\LoggerAwareTrait;

    class MyClass implements SingletonInterface, LoggerAwareInterface {
       use LoggerAwareTrait;

        ...
           $this->logger->info('My class is executed.');
           if ($error) {
              $this->logger->error('error in class MyClass');
           }

>         ...

Reviewer 2: I would have preferred that the introduction of SingletonInterface get's another issue.
Anyway, now the namespace for the SingletonInterface is missing.

Reviewer 2: lines 185 + 210
